Feature guided fill

AndyLV opened the pull request:
As asked for in https://github.com/inkstitch/inkstitch/issues/1196#issuecomment-1008246658 here is the corresponding pull request.
The _feature_guided_fill_ includes the parts discussed here: https://github.com/inkstitch/inkstitch/issues/1196 and here: https://github.com/inkstitch/inkstitch/issues/1359

kaalleen commented:
Thanks :)

kaalleen commented:
I believe the changes e.g. in base.py are not meant to be made, but are caused by changes in the main branch while you were working on yours only locally?

I am wondering how we are going to work on this branch. Would you rather have us to upload changes directly to your branch (talking with code) or do you want us to comment on the branch, so you have full control of it?

AndyLV commented:
Yes, this is correct - there are some changes caused by progress in the main branch while I was working locally. My original plan was to merge the meanwhile progress with my branch as soon as the new GUI is available (mentioned in https://github.com/inkstitch/inkstitch/issues/1196#issuecomment-899891183).

I do not need to have full control of the code - feel free to modify it as you like.

kaalleen commented:
Ok, great. The params branch is going to take much longer than we originally planed to. So we can push this one forward in the meantime. If you don't mind... this would be a great time then to merge the changes into here, so we have it up to date.

I am curious though, what the reasons are for the removal of the old fill method (fill.py). We could make this feature only available for autofill. But also keep the other fill method for backwards compatibility.

AndyLV commented:
I merged my code with the recent main branch. In addition, I re-added the legacy fill method again. Before completing the pull request please check that your features are all working since there is a lot of code changing with this pull request. In addition the translations need to be updated.

kaalleen commented:
Thank you! The translations (as long as they are marked as translatable strings _()) will automatically be transfered to crowdin, once this branch is merged into main. So we don't have to worry about them.
I pushed some changes to this branch. First I wanted to remove some too complex warnings and very few other style issues. We still want to check, if it isn't possible to make those complex functions less complex though. Also I tried to "untangle" the pattern and marker mixup, which better should be separated.

AndyLV commented:
Thank you - your modifications to the naming and structure of "marker" and "pattern" makes the code better understandable.

kaalleen commented:
Again I pushed some changes. I hope I didn't break anything with it :D 

* I renamed autofill to fill_stitch (since this was planed in the params to electron branch)
* I split the to_stitch_group function in fill_stitch, so we don't have the too complex warning and don't need to suppress it
* If there is no guide line for the guided fill, we can use a fallback to the normal autofill functionality - and give out a warning that this will happen in the troubleshoot function. This way, we can avoid an error message in the first place. What do you think?
* I updated the symbol for guide lines, so people will not confuse it with the pattern symbol
* Do not render guide line objects. We can treat them just like patterns. Therefore I renamed the pattern object to a marker object. I think (?!) we don't need to be so specific at that point. This will help us to inform the users about the fact, that this object will not be rendered directly in the troubleshoot section - and how to remove the marker if they want to

Well, I think that's about it. Hope you agree with the changes.

AndyLV commented:
> I renamed autofill to fill_stitch (since this was planed in the params to electron branch)

I think fill_stitch sounds more general which is better since we have different fill options available now.

>    If there is no guide line for the guided fill, we can use a fallback to the normal autofill functionality - and give out a warning about this behaviour in the troubleshoot function. This way, we can avoid an error message in the first place. What do you think?

On my PC the new code crashed when I used guided fill without having created a guide line. I pushed a fix - unfortunately my automatic formatting tool introduced several new changes - the only relevant is in line 602.

>    I updated the symbol for guide lines, so people will not confuse it with the pattern symbol

Looks also more intuitive, thank you.

>    Do not render guide line objects. We can treat them just like patterns. Therefore I renamed the pattern object to a marker object. I think (?!) we don't need to be so specific at that point. This will help us to inform the users about the fact, that this object will not be rendered directly in the troubleshoot section - and how to remove the marker if they want to

I'm sorry - what do you mean with render objects (do you mean the preview)? At least I did not notice that an object without params settings was rendered in the preview so far?

kaalleen commented:
> I'm sorry - what do you mean with render objects (do you mean the preview)? At least I did not notice that an object without params settings was rendered in the preview so far?

Yes, I saw them being rendered. But that's ok now. No worries.

Thanks for the fix. I did not notice it along the way.

Is it possible to set your automatic formatting tool to a linelength of 150? That's what we use in our style check.

I am quite busy at the moment and a bit stressed. So I am not very well concentrated and I am not sure, when I'll find the time to work on this branch. If you want to push it forward, there is maybe something you can do in the meantime...

Since we worked ouselves through the elements section, we next want to have a closer look at the stitches section.

You did add some new files in there. We usually do not use snake case for file names. It'd be nice, if they were written in lower case letters with underscores. Maybe we can also make sure, that people will understand directly to which embroidery style they belong to. At least the StitchPattern.py file, could be named tangential_fill.py (if that does make sense?).

When I opened one of the files, I could see a lot of print statements in there (btw. I speak german too ;). They maybe could be cleaned up a little?

Thanks again for all the work you've put into this. It will make a lot of people happy :)

AndyLV commented:
I cleaned the code up and changed the naming of some files to match your style guide. 

Do you have an idea why the windows build fails with error messages like "info There appears to be trouble with your network connection. Retrying..."?

kaalleen commented:
Thank you :) I'll go try to go on soon.

The windows build failure has nothing to do with the code. It is actually what it says... the network connection failed and it couldn't finish the build (download the Ink/Stitch dependencies). If we'd rerun the build it should work. Lately we saw this issue quite often.

AndyLV commented:
I agree that it might be easier to read if autofill and guided_fill are separated. However, they share the main routing idea and especially the routines _build_guided_fill_stitch_graph_ &  _build_fill_stitch_graph_ as well as  _process_travel_edges_ (in autofill and guided_fill) share a lot of code. Maybe it is better to use the same routine for _build_fill_stitch_graph_ and _process_travel_edges_ to avoid too much code repetition (--> https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)? Especially if one finds a bug or applies some changes to one of this routines in e.g. autofill it is very likely that this change is also required for guided_fill but it might easily be forgotten in this duplicate code.

kaalleen commented:
I suppose you are right about the mentioned methods and actually was waiting for some protest. We can move those back later on. Sorry for that.

I am looking through the code here and there and I am wondering if the options "interlaced" and "flip" really should be options for the guided fill? The flip option applies usually to the legacy fill, which isn't cannot be guided by start- and end symbols. I also couldn't make out any difference in guided fill if the flip option is enabled or not - while it follows the start- and end commands as it should.
The interlaced option - I may be wrong - seems to be part of tangential fill more than the guided fill?

In repair_multiple_parallel_offset_curves (guided_fill.py) it looks like the linemerge is overwritten by the very next line?

AndyLV commented:
> I suppose you are right about the mentioned methods and actually was waiting for some protest. We can move those back later on. Sorry for that.

Don't worry - I'm not a professional programmer so avoiding duplicates was only a suggestion.

>I am looking through the code here and there and I am wondering if the options "interlaced" and "flip" really should be options for the guided fill? The flip option applies usually to the legacy fill, which cannot be guided by start- and end symbols. I also couldn't make out any difference in guided fill if the flip option is enabled or not - while it follows the start- and end commands as it should.

>The interlaced option - I may be wrong - seems to be part of tangential fill more than the guided fill?

"flip" is only present since it was in the original autofill code so I only kept it - if you say we can remove it is fine for me.

Regarding "interlaced" the situation is a little bit more complicated. You are right that the code comes originally from the tangential fill method. According to my understanding there are currently two completely different point placement strategies:
-Auto_fill places the points equidistant along the stitching direction with configurable offsets between adjacent lines
-Tangential fill can not place all points equidistant since in curved line segments  smaller point distances are required than in flat line segments. Hence it needs to adjust the point spacing dynamically to avoid that features in the original curve disappear due to too large stitching distances. However, if the points are only placed according to the constraint that all relevant features shall be present in the stitched result, the result might not look very well because there will be nearly no correlation between the points placed in adjacent offsetted curves. That is why I introduced the parameter "interlaced" - if true, the code tries to place a point in the middle of two points of adjacent curves. If false, the code tries to place a point close to one of an adjacent curve. This parameter is a little bit like the "stagger" parameter - however, currently it is only limited to stagger=1 (interlaced=false) and stagger=2 (interlaced=true). 

Coming back to guided fill I needed to select one of this two strategies. Since the user created guide line has no hard constraints (e.g. regarding curvature), the first method might not be usable in situations with large curvature (e.g. a zig-zag line). Hence I selected the second strategy which comes with the parameter "interlaced".

>In repair_multiple_parallel_offset_curves (guided_fill.py) it looks like the linemerge is overwritten by the very next line?

Yes you are right - thank you for the hint - it should be:
    lines = list(lines.geoms)

lexelby commented:
I've been away for quite awhile, and so I've only just now tried out this branch for the first time.  I am stunned.  Just floored.  The results you're getting are nothing short of amazing!  I can't wait to get this in front of our users.  So much awesome work here!

kaalleen commented:
Thank you for your detailed explaination.

I am currently not able to find any time to work on Ink/Stitch. I'll have to step back for a while, sorry.

claudinepeyrat06 commented:
Really really great feature.  Thank you so much.
Just tried to embroider a spiral fill in a circle. Do you think that it would be possible to  control both maximum and minimum length of stitches ? I ended up with very small stitches near the center and my embroidery machine does not like it very much. Or may be there is something that I did not understand....
I love the guided line feature, so easy to use, and so useful. Thanks again.

AndyLV commented:
@kaalleen Thank you very much for your support so far :-)

@claudinepeyrat06 I introduced a minimum stitch length parameter. Note that using this parameter makes the point placing pattern less regular (as soon as this parameter comes into play) and that with increasing value the deviation to the intended shape increases.

danielkschneider commented:
Really two nice and powerful new features ! Thanx very much :)

I did a test on both guided fills and tangential fills. I noticed that tangential doesn't work work with certain shapes. Just saying in case you are not aware, not complaining.... 
![2022-03-20 20_59_58-tangential-fill svg - Inkscape](https://user-images.githubusercontent.com/2872643/159183887-b64f3f18-0666-448c-8666-2b14259b14ac.png)

![tangential-fill](https://user-images.githubusercontent.com/2872643/159183941-dee8873f-b090-4341-ac50-9aaac1b3224f.svg)

lexelby commented:
I've started digging into the code, making slow progress.  Just wanted to let you know I'm looking at it.

AndyLV commented:
@danielkschneider Thank you very much for these examples. It showed me some unrecognized problems with Shapelys parallel_offset routine that is why I followed lexelbys advice to use the buffer routine instead. Your examples should work, now.

danielkschneider commented:
> @danielkschneider Thank you very much for these examples. It showed me some unrecognized problems with Shapelys parallel_offset routine that is why I followed lexelbys advice to use the buffer routine instead. Your examples should work, now.

cool 👍

danielkschneider commented:
> @danielkschneider Thank you very much for these examples. It showed me some unrecognized problems with Shapelys parallel_offset routine that is why I followed lexelbys advice to use the buffer routine instead. Your examples should work, now.

```
sudo dpkg -i inkstitch_0.0.1-feature_guided_fill_amd64.deb 
dpkg: error processing archive inkstitch_0.0.1-feature_guided_fill_amd64.deb (--install):
 parsing file '/var/lib/dpkg/tmp.ci/control' near line 2 package 'inkstitch':
 'Version' field value '0.0.1-feature_guided_fill': invalid character in revision number
Errors were encountered while processing:
 inkstitch_0.0.1-feature_guided_fill_amd64.deb
```

```
sb_release -a
LSB Version:	core-11.1.0ubuntu2-noarch:security-11.1.0ubuntu2-noarch
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.4 LTS
Release:	20.04
Codename:	focal
```

lexelby commented:
@danielkschneider That commit should fix your deb bug.

lexelby commented:
@AndyLV Okay!  I've finally had a chance to read through the code and gain an understanding of what's going on here.  I've got a lot of feedback/ideas.

First, what I love:
* The spiral strategy is awesome.  I love the idea of interpolating between the concentric paths.
* The general concept of doing successive buffers, building a tree, and stitching them together in various ways is really nifty.
* The results are really cool and I can't wait to get this shipped.

Now, some areas I want to improve.  The biggest thing is that this code just takes a long time to run.  It's pretty much untenable, I'm afraid.  There are areas where we can optimize certain code, but I think in general we're going to need to consider some new approaches to make this workable.

For example, in sample_linestring, the main loop does repeated calls to `simplify()`, and this is just going to take a long time, period.  Between that and the whole point-transfer strategy, we're spending a _ton_ of time figuring out where to place the stitches along the offset paths.

I'm actually not all that keen on the way the stitches are placed, for that matter.  The stitches end up lining up in weird ways that I think will stand out.  In general, this is a hard problem, way harder than simple fill with parallel lines of stitches, and I see exactly why you've gone in the direction you did.  I have some ideas for other ways we could go about this, and I'll elaborate in a later comment.  If we can find a simpler and more efficient way to place the stitches, it would make the code dramatically simpler and cut the running time down a ton.

Another thing that I suspect the code spends a lot of time on is calculating nearest points for the "closest point" strategy.  I'm not sure I understand the benefit here, TBH.  `buffer()` is specifically trying to find a path that is offset equidistantly from the path you give it.  By definition, the two curves will be _almost_ the same distance from each other everywhere, and `nearest_points()` will produce an effectively random result (dictated primarily by imprecision in the paths).  And it'll spend a _lot_ of time to find those random points.  I actually don't really see the benefit to this strategy at all; the path it produces seems to go all over the place.  Maybe we should get rid of it?

For the "single spiral" strategy, I like the algorithm in `interpolate_LinearRings()` -- I never would have thought of doing it that way.  Neat!  I think our use case might let us simplify the code and make it more efficient at the same time.  Building a KD tree may not be necessary.  Our two paths will be very similar; they're not just two arbitrary paths with no relation to each other.  What if we use `resample_path()` on both with the same `count` (instead of passing a `step`)?  Then we could directly interpolate between the corresponding points on each resampled line.

In general, I love what we're doing here.  There are some places in the code where I'd like to make a few changes beyond the above, mostly for clarity and consistency.  I know it's hard to simplify code when what you're doing is complex, but I'd like to try to avoid `noqa: C901` as much as possible.  Fortunately, I think removing the point-transfer strategy will go a long way toward that goal.

For the rest, I'll add comments inline as a pull request review.

lexelby commented:
For stitch placement, I was wondering if we could do something like the current "stagger" system.  The stagger code does a lot of calculating, but in the end, it's effectively just intersecting the row with a grating that's at a shallow angle.  What if we do something similar for tangential and guided fill?

![image](https://user-images.githubusercontent.com/4446653/161663935-1077b5bd-3e72-4c56-ae43-27a8d3c2ecad.png)

This actually kinda works:

![277589138_876890313142772_7970848244337538275_n](https://user-images.githubusercontent.com/4446653/161664067-809457f8-b63a-4ad6-81e4-e50e008932b9.png)

But it fails toward the top and bottom where the individual stitches come too far apart:

![277824051_367184428624946_8800516853770594723_n](https://user-images.githubusercontent.com/4446653/161664152-0328ba8a-32e3-4338-8d68-ebbba5714292.png)

(I was just simulating my idea in inkscape, so there are some extra nodes, which you can ignore).

So it's not quite right, but it gets us pretty close.  I think if we broke up those long sections just by splitting them up evenly or every max_stitch_len, we'd end up with a workable solution.

The only thing I fear is that the diagonal stagger pattern much drown out the spiral pattern.  My hunch is that it won't, though.  I don't have access to an embroidery machine just now, so I can't test it.

lexelby commented on lib/stitches/guided_fill.py:
```diff
@@ -0,0 +1,355 @@
+def stitch_line(stitches, stitching_direction, geometry, projected_points, max_stitch_length, min_stitch_length, row_spacing, skip_last, offset_by_half):
+    if stitching_direction == 1:
+        stitched_line, _ = raster_line_string_with_priority_points(
+            geometry, 0.0, geometry.length, max_stitch_length, min_stitch_length, projected_points, abs(row_spacing), offset_by_half, True)
```
I'm pretty sure this `abs()` isn't needed.  I don't see how `row_spacing` could be negative.  Same for the one a couple lines down.

lexelby commented on lib/stitches/guided_fill.py:
```diff
@@ -0,0 +1,355 @@
+    return segments
+
+
+def process_travel_edges(graph, fill_stitch_graph, shape, travel_edges):
```
I see what you're trying to do by defining get_segments) and process_travel_edges() here, but it doesn't work.  These ones here are never run.  `guided_fill()` calls `build_travel_graph()` from auto_fill.py, and `build_travel_graph()` will call the `process_travel_edges()` that's in auto_fill.py.

lexelby commented on lib/stitches/guided_fill.py:
```diff
@@ -0,0 +1,355 @@
+
+    for edge in path:
+        if edge.is_segment():
+            new_stitches = []
+            current_edge = fill_stitch_graph[edge[0]][edge[-1]]['segment']
+            path_geometry = current_edge['geometry']
+            projected_points = current_edge['projected_points']
+            stitching_direction = 1
+            if (abs(edge[0][0]-path_geometry.coords[0][0])+abs(edge[0][1]-path_geometry.coords[0][1]) >
+                    abs(edge[0][0]-path_geometry.coords[-1][0])+abs(edge[0][1]-path_geometry.coords[-1][1])):
+                stitching_direction = -1
+            stitch_line(new_stitches, stitching_direction, path_geometry, projected_points,
+                        max_stitch_length, min_stitch_length, row_spacing, skip_last, offset_by_half)
+            current_edge['already_rastered'] = True
+            transfer_points_to_surrounding_graph(
+                fill_stitch_graph, current_edge, row_spacing, False, new_stitches, overnext_neighbor=True)
+            transfer_points_to_surrounding_graph(fill_stitch_graph, current_edge, row_spacing, offset_by_half,
+                                                 new_stitches, overnext_neighbor=False, transfer_forbidden_points=offset_by_half)
+
+            stitches.extend(new_stitches)
```
There's a critical piece missing here.  After stitching a line, we need to remove all of the edges in the travel graph that the stitched line touched.  Without that, travel stitches can go on top of lines we've already stitched -- I've seen it happen in my tests.

lexelby commented on lib/stitches/tangential_fill_stitch_pattern_creator.py:
```diff
@@ -0,0 +1,912 @@
+    return result_coords, result_coords_origin
+
+
+# Partly taken from https://github.com/mikedh/pocketing/blob/master/pocketing/polygons.py
```
In order to use the code from pocketing like this, we probably have to include its BSD license and copyright statement.  I'd like to avoid doing that if possible because it complicates our own license, which is GPL.

Fortunately, I don't think we'll need to worry about that if we implement my idea to get rid of the KD tree parts, since we'll end up with a function that is substantially different.  At that point we'll just be using pocketing's code as inspiration rather than directly copying.

lexelby commented:
I'm hoping I might be able to work on some or all of what I've discussed over the next couple days -- we'll see.

AndyLV commented:
I'm sorry - I've currently only limited time so only a short comment:

According to my understanding you have two simple possibilities to realize the stagger system:
1) Taking only the stagger points + sub-points which were introduced to not exceed the maximum stitch distance
2) Taking stagger points + sub-points + all points which make up the shape on its own

Point 1) has the disadvantage that you might miss sharp or important edges causing noticeable deviations from the intended shape to stitch (e.g. stitch a star)
Point 2) does not have this disadvantage but might have a lot of extra points which are not really necessary

When I implemented the tangential_fill version I faced a similar problem:
On one hand I've the projected points from the already stitched adjacent shapes - however, these points might not be enough to guarantee a small deviation to the intended shape. Unfortunately the only solution which came to my mind was this computationally expensive "simplify" code which you mention above. The idea was the following:
I merge the projected points with the all points which make up the shape on its own (similar to 2)). Then I walk along the path and leave out not necessary points of the original shape as long as the used projected points do not introduce a too strong deviation to the full shape (whether points can be left out is determined whether simplify would remove them). If necessary (e.g. at hard edges) I use points of the original shape next to the projected points to minimize the deviation to the desired shape.

I agree that this method is quite computationally expensive - that is why I asked for a progress bar (https://github.com/inkstitch/inkstitch/issues/1196#issuecomment-902797352)

lexelby commented:
Ah yes, I forgot to mention that part, thanks.  I was thinking something along the line of #2.  The idea would be to add stitches at corners with an angle sharper than a certain threshold.  I like our code for calculating all angles at once using numpy, hopefully that should run pretty fast.

One thing to think about as we do this is that we have a fair bit of leeway.  Embroidery machines operate with a resolution of 0.1mm, but in reality we can be further off than that without making a distinguishable difference.  If the user doesn't like it, they can decrees the stitch length.

lexelby commented:
This latest commit refactors things to avoid needing `anytree`.  It bugged me that `anytree` contains a couple thousand lines of code, but the same functionality is largely contained in networkx.  I wanted to keep it to just one graph library.

'Maybe overkill, but this will also reduce our packaged application size and it's one less dependency we have to keep track of.

AndyLV commented:
> Another thing that I suspect the code spends a lot of time on is calculating nearest points for the "closest point" strategy. I'm not sure I understand the benefit here, TBH. buffer() is specifically trying to find a path that is offset equidistantly from the path you give it. By definition, the two curves will be almost the same distance from each other everywhere, and nearest_points() will produce an effectively random result (dictated primarily by imprecision in the paths). And it'll spend a lot of time to find those random points. I actually don't really see the benefit to this strategy at all; the path it produces seems to go all over the place. Maybe we should get rid of it?

The original idea behind it was to avoid choosing a large distances by taking a point at a (sharp) edge where a multiple of the original offset value can occur. I agree that this goal can be achieved cheaper - I did not change it so far since I liked the resulting "random" pattern which avoids an accumulation of the connecting elements like in the "inner to outer" method. In addition I thought that the calculation time required for the closest point is small compared to the point placing routines.

>What is this modulo for? The if condition means that the modulo operator should do nothing, I think?

To be honest - I'm not fully sure why I placed a modulo operator there. One reason might be my uncertainty about the evaluation order - if the "if" conditions is evaluated after having accessed the array element one might get an out-of-range error.

>I can't find a way that row_spacing could be negative...
I'm pretty sure this abs() isn't needed. I don't see how row_spacing could be negative. Same for the one a couple lines down.

I think this has historical reasons since I worked in the beginning also with negative offsets and that is why I usually calculated the abs value in situations where I do not want a signed offset. However, I agree - the abs is probably dispensable.


>I see what you're trying to do by defining get_segments) and process_travel_edges() here, but it doesn't work. These ones here are never run. guided_fill() calls build_travel_graph() from auto_fill.py, and build_travel_graph() will call the process_travel_edges() that's in auto_fill.py.

There was a discussion about the doubled code (see https://github.com/inkstitch/inkstitch/pull/1548#issuecomment-1044999971). I was not so happy with the code doubling but I didn't wanted to undo the work of kalleen since I'm very glad about her support and I do not claim that my view is correct. Nevertheless we should fix what you found out (I did not notice it before).

>There's a critical piece missing here. After stitching a line, we need to remove all of the edges in the travel graph that the stitched line touched. Without that, travel stitches can go on top of lines we've already stitched -- I've seen it happen in my tests.

You are absolutely right - I remember that I wanted to introduce this part in guided fill but it seems that I've forgotten it - good that you noticed it.

>Okay, here's a new version of interpolate_linear_rings() that works as I suggested above, and it's way faster. I did have to resample the rings with quite a few points in order to avoid rounding corners, but numpy makes it still quite a lot faster than building the KD tree.

Looks good - my only concern would be that the method might fail if there are several edges close to the origin? However, this would probably happen only in very rare cases.

Regarding the performance of the point-transfer routines: The most time is spend on calculating the intersection of normal vectors with the next LinearRing. Hence I thought about testing "prep" https://shapely.readthedocs.io/en/stable/manual.html#prepared-geometry-operations in the past but I currently do not find the time for it.

All in all you have improved the code a lot (e.g. also the angle calculations, etc.) :-)

lexelby commented:
Okay!  I did a pretty major rework and refactor.  As mentioned above, I got rid of `point_transfer` and `sample_linestring`. The results weren't good enough to justify the tens-of-seconds runtime.  I replaced them with running stitch.  I'd be happy to revisit the idea of separating neighboring points, but we'll need to find a way to do this much more efficiently for it to be feasible.

I rewrote `running_stitch()`, and it works a ton better now!  Previously it (and the tangential fill code) tried to find corners by calculating the angles formed by successive line segments.  I realized that this approach is fundamentally flawed.  A sharp corner can be made of many small line segments, each with a relatively shallow angle, and it can still look to a human as if it's a corner.  This can happen even if you `simplify()` first.

Instead, I used `simplify()` to find the corners for us.  It does a great job of removing any points that aren't necessary.  My algorithm stitches all of the points that `simplify()` keeps, and then adds more stitches between them in order to satisfy the stitch length.  I've already used this running stitch algorithm on a patch I was designing and the results are really promising!

So now, everything uses running_stitch.  This does mean that we can have stitches in parallel curves line up, leading to potentially distracting lines.  In my tests, I saw that this really only shows up if you try to do a really dense fill.  I'm not actually convinced that people will use tangential fill for that.  I found that it didn't do a very good job of actually filling a shape completely, like normal fill stitch.

Guided fill is a different story -- we'll see if maybe we need to come up with a better solution.  Fundamentally, I don't think that a perfect solution is possible, since each row is shaped differently from the previous one.

I added a couple of new features while I was working on this.  In Inner to Outer mode, I added an "avoid self-crossing" option:

![image](https://user-images.githubusercontent.com/4446653/166614396-8046b53d-64bc-4891-838c-647c0e347107.png)

That was surprisingly easy to implement and gives a really nice result!

For Single Spiral mode, I added a new mode, Double Spiral.  The code is almost the same as Single Spiral mode, but it does a [Fermat Spiral](https://en.wikipedia.org/wiki/Fermat's_spiral).

This is almost ready to merge.   I'm thinking of renaming "tangential fill" to "contour fill".  This term will probably be more familiar to English speakers and give more of a clue about what it does.

kaalleen commented:
Rebased the branch to main to pull in the changes.
Note that there has been an update to the inkex version.

lexelby commented:
Oh!  GitHub didn't show me your comment before I made mine above.  Sorry about that.


> The original idea behind it was to avoid choosing a large distances by taking a point at a (sharp) edge where a multiple of the original offset value can occur. I agree that this goal can be achieved cheaper - I did not change it so far since I liked the resulting "random" pattern which avoids an accumulation of the connecting elements like in the "inner to outer" method. In addition I thought that the calculation time required for the closest point is small compared to the point placing routines.

Ah, thanks!  I appreciate the context here.

I tried changing `_get_nearest_points_closer_than_thresh` to choose a random point, as opposed to the first point within the threshold, and it produces a nifty random pattern much like closest_neighbor did.  It looks especially cool with my new "avoid self-crossing" feature.  I might add this as an option at some point.

> There was a discussion about the doubled code (see [#1548 (comment)](https://github.com/inkstitch/inkstitch/pull/1548#issuecomment-1044999971)). I was not so happy with the code doubling but I didn't wanted to undo the work of kalleen since I'm very glad about her support and I do not claim that my view is correct. Nevertheless we should fix what you found out (I did not notice it before).

I managed to avoid pretty much all of the code duplication in this latest version, with only just a little refactoring of the existing autofill code.

> Looks good - my only concern would be that the method might fail if there are several edges close to the origin? However, this would probably happen only in very rare cases.

By origin, do you mean the start of one of the rings?  I considered seeing if I could add the vertices of both of the rings as interpolation points, which might avoid the need to resample to 20 points per stitch, but I couldn't figure out how to do it...  I'll probably revisit at some point.

> Regarding the performance of the point-transfer routines: The most time is spend on calculating the intersection of normal vectors with the next LinearRing. Hence I thought about testing "prep" https://shapely.readthedocs.io/en/stable/manual.html#prepared-geometry-operations in the past but I currently do not find the time for it.

Ah, okay.  I'll be honest: I never fully read through the point-transfer stuff, the code was very dense and I kept getting lost.  Perhaps we should revisit at some point.  It might be that we can handle this in running_stitch, simply by keeping a running list of stitches placed so far, and use that to find a good spot for the first stitch in each section.  An STRTree might make this reasonably efficient.  Hmmm.... probably best to save this for a later PR.

> All in all you have improved the code a lot (e.g. also the angle calculations, etc.) :-)

I did end up getting rid of that code, but I came across an interesting technique you might like.  Instead of using the dot product and arc cosine, you can use the arc tangent.  You use np.diff() to create an array containing vectors, each going from one point in the linestring to the next.  Then you use arctan2(y, x) and that gives you the angle of that segment with respect to the X axis.  After that, you subtract each angle from the next (that's np.diff() again) to find out the angle between each segment.  Along with being more efficient (no need to calculate vector lengths), arctan2 can handle all angles unambiguously, unlike arc cosine.

lexelby commented:
Okay, I renamed "tangential fill" to "contour fill".  I also cleaned up some params we're no longer using and fixed a couple bugs I'd introduced.

I think this is pretty close to ready to go.

lexelby commented:
@kaalleen and I discussed and decided that having single and double spiral throw an exception isn't great for the user.  There's no information on what makes a shape good for a spiral.  Instead, we're going to do a "best effort", where it just tries as well as it can to make a single spiral.  That way the user has more of a chance to guess what might be wrong and adjust their shape to make it come out right.  Here's an example:

![image](https://user-images.githubusercontent.com/4446653/167231105-cdb2d415-95b7-4016-bbfc-e335d8f67b78.png)

Obviously this looks bizarre, and it even cuts through the holes, but maybe it nudges the user in the right direction.

lexelby commented:
In some cases it may well even be exactly what the user wants: 
![image](https://user-images.githubusercontent.com/4446653/167231374-918a7e09-ea3e-429a-ab3f-e39f6dbab002.png)

lexelby commented:
Okay!  I think this thing is ready for a final review and then we can merge :smile:

kaalleen commented:
@AndyLV are you ok with the branch as is?

AndyLV commented:
> @AndyLV are you ok with the branch as is?

I was a little bit unhappy that everything related to stitch point placing method was removed. I had preferred to allow the user to select between both methods (offering more design options) since this would also have allowed to improve the too computational expensive method more easily.
Nevertheless, from my side the current state of the branch is ready to be merged.

lexelby commented:
@AndyLV Thanks for your flexibility, and I'm sorry this isn't quite what you wanted.  I do think there's lots of room for improvement of the stitch placement, and we can work on that in later versions.  The code is still there in the Git history, after all. :)  For now, let's go ahead and ship this as it is and see what the users think.
